Sources/CardPayments/CardClient.swift

```diff
@@ -7,27 +7,87 @@ import CorePayments
+    
     private let apiClient: APIClient
     private let config: CoreConfig
     private let webAuthenticationSession: WebAuthenticationSession
     private var analyticsService: AnalyticsService?
+    private var graphQLClient: GraphQLClient?
 
     /// Initialize a CardClient to process card payment
     /// - Parameter config: The CoreConfig object
     public init(config: CoreConfig) {
         self.config = config
         self.apiClient = APIClient(coreConfig: config)
         self.webAuthenticationSession = WebAuthenticationSession()
+        self.graphQLClient = GraphQLClient(environment: config.environment)
     }
 
     /// For internal use for testing/mocking purpose
-    init(config: CoreConfig, apiClient: APIClient, webAuthenticationSession: WebAuthenticationSession) {
+    init(
+        config: CoreConfig,
+        apiClient: APIClient,
+        webAuthenticationSession: WebAuthenticationSession,
+        graphQLClient: GraphQLClient? = nil
+    ) {
         self.config = config
         self.apiClient = apiClient
         self.webAuthenticationSession = webAuthenticationSession
+        self.graphQLClient = graphQLClient
     }
-
+    
+    public func vault(vaultRequest: CardVaultRequest) {
+        Task {
+            guard graphQLClient != nil else {
+                // TODO: this will be APIClient after networking layer refactor
+                throw CardClientError.unknownError
+            }
+            do {
+                let card = vaultRequest.card
+                let setupTokenID = vaultRequest.setupTokenID
+                let (updateResult) = try await updateSetupToken(vaultSetupTokenID: setupTokenID, card: card)
+                if let result = updateResult {
+                    // can it be not approved and end up here?
+                    // TODO: handle 3DS contingency with helios link
+                    if let link = result.links.first(where: { $0.rel == "approve" && $0.href.contains("helios") }) {
+                        let url = link.href
+                        print("3DS url \(url)")
+                    } else {
+                        let vaultResult = CardVaultResult(setupTokenID: result.id, status: result.status)
+                        notifyVaultSuccess(for: vaultResult)
+                    }
+                } else {
+                    notifyVaultFailure(with: CardClientError.noVaultTokenDataError)
+                }
+            } catch  let error as CoreSDKError {
+                // TODO: right now, GraphQLError is thrown, but will need to add new errors after refactor
+                notifyVaultFailure(with: error)
+            } catch {
+                notifyVaultFailure(with: CardClientError.vaultTokenError)
+            }
+        }
+    }
+    
+    func updateSetupToken(
+        vaultSetupTokenID: String,
+        card: Card
+    ) async throws -> TokenDetails? {
+        guard let graphQLClient else {
+            throw CardClientError.unknownError
```

Comment:
Should we add a more specific error for this?

Reply:
I thought about it but I think we are not going to be using graphQLClient after refactor so I left it that way.